src/System.Linq/src/System/Linq/Distinct.cs

```diff
@@ -116,7 +116,7 @@ public List<TSource> ToList()
                 return FillSet().ToList();
             }
 
-            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer) => new HashSet<TSource>(this, comparer);
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer) => _source.ToHashSet(comparer);
```

Comment:
This is incorrect if the Distinct iterator has a different comparer. (Also, it would be good to add a regression test for this.)